src/storage/src/hummock/block_cache.rs

```diff
@@ -12,56 +12,117 @@
+        }
+    }
+}
+
+impl AsRef<Block> for BlockHolder {
+    fn as_ref(&self) -> &Block {
+        unsafe { &(*self.block) }
+    }
+}
+
+unsafe impl Send for BlockHolder {}
+unsafe impl Sync for BlockHolder {}
 
 pub struct BlockCache {
-    inner: Cache<Bytes, Arc<Block>>,
+    inner: Arc<LruCache<(u64, u64), Box<Block>>>,
 }
 
 impl BlockCache {
     pub fn new(capacity: usize) -> Self {
-        let cache: Cache<Bytes, Arc<Block>> = Cache::builder()
-            .weigher(|_k, v: &Arc<Block>| v.len() as u32)
-            .initial_capacity(capacity / DEFAULT_ENTRY_SIZE)
-            .max_capacity(capacity as u64)
-            .build();
-        Self { inner: cache }
+        let cache = LruCache::new(CACHE_SHARD_BITS, capacity, DEFAULT_OBJECT_POOL_SIZE);
+        Self {
+            inner: Arc::new(cache),
+        }
     }
 
-    // TODO: Optimize for concurrent get https://github.com/singularity-data/risingwave/pull/627#discussion_r817354730.
-    pub fn get(&self, sst_id: u64, block_idx: u64) -> Option<Arc<Block>> {
-        self.inner.get(&Self::key(sst_id, block_idx))
+    pub fn get(&self, sst_id: u64, block_idx: u64) -> Option<BlockHolder> {
+        self.inner
+            .lookup(Self::hash(sst_id, block_idx), &(sst_id, block_idx))
+            .map(BlockHolder::from_cached_block)
     }
 
-    pub async fn insert(&self, sst_id: u64, block_idx: u64, block: Arc<Block>) {
-        self.inner.insert(Self::key(sst_id, block_idx), block).await
+    pub fn insert(&self, sst_id: u64, block_idx: u64, block: Box<Block>) {
+        self.inner.insert(
+            (sst_id, block_idx),
+            Self::hash(sst_id, block_idx),
+            block.len(),
+            block,
+        );
     }
 
     pub async fn get_or_insert_with<F>(
         &self,
         sst_id: u64,
         block_idx: u64,
         f: F,
-    ) -> HummockResult<Arc<Block>>
+    ) -> HummockResult<BlockHolder>
     where
-        F: Future<Output = HummockResult<Arc<Block>>>,
+        F: Future<Output = HummockResult<Box<Block>>>,
     {
-        self.inner
-            .try_get_with(Self::key(sst_id, block_idx), f)
-            .await
-            .map_err(HummockError::other)
+        let h = Self::hash(sst_id, block_idx);
+        let key = (sst_id, block_idx);
+        match self.inner.lookup_for_request(h, key) {
+            LookupResult::Cached(entry) => Ok(BlockHolder::from_cached_block(entry)),
+            LookupResult::WaitPendingRequest(recv) => {
+                let entry = recv.await.map_err(HummockError::other)?;
+                Ok(BlockHolder::from_cached_block(entry))
+            }
+            LookupResult::Miss => match f.await {
+                Ok(block) => {
+                    let entry = self.inner.insert(key, h, block.len(), block);
+                    Ok(BlockHolder::from_cached_block(entry))
+                }
+                Err(e) => {
+                    self.inner.clear_pending_request(&key, h);
+                    Err(e)
```

Comment:
Does it mean if we have multiple requests and the first request fails, all remaining requests will fail?

Reply:
yes